### PR TITLE
separate HTTP client pool for cancellation requests

### PR DESCRIPTION
**Pull request by xvrl**
* reduces contention between queries and cancellation requests
* more aggressive timeouts for cancellation requests

### CODE REVIEW

File: server/src/main/java/io/druid/server/AsyncQueryForwardingServlet.java [hunk 1]
```diff
@@ -110,6 +114,12 @@ public AsyncQueryForwardingServlet(
     this.httpClientConfig = httpClientConfig;
     this.emitter = emitter;
     this.requestLogger = requestLogger;
+
+    // separate client with more aggressive connection timeouts
+    // to prevent cancellations requests from blocking queries
+    this.broadcastClient = httpClientProvider.get();
+    broadcastClient.setConnectTimeout(CANCELLATION_TIMEOUT);
+    broadcastClient.setMaxRequestsQueuedPerDestination(MAX_QUEUED_CANCELLATIONS);
```
Comment (nishantmonu51) [inline server/src/main/java/io/druid/server/AsyncQueryForwardingServlet.java]: should we also lower down maxConnectionsPerDestination  ?

File: server/src/main/java/io/druid/server/AsyncQueryForwardingServlet.java [hunk 2]
```diff
@@ -68,6 +68,8 @@
   private static final String HOST_ATTRIBUTE = "io.druid.proxy.to.host";
   private static final String QUERY_ATTRIBUTE = "io.druid.proxy.query";
   private static final String OBJECTMAPPER_ATTRIBUTE = "io.druid.proxy.objectMapper";
+  public static final int CANCELLATION_TIMEOUT = 500;
+  public static final int MAX_QUEUED_CANCELLATIONS = 64;
```
Comment (nishantmonu51) [inline server/src/main/java/io/druid/server/AsyncQueryForwardingServlet.java]: isn't this probably too low ? 
can we make it configurable via runtime.props?

File: server/src/main/java/io/druid/server/AsyncQueryForwardingServlet.java (hunk 1 above)
Comment (xvrl) [inline server/src/main/java/io/druid/server/AsyncQueryForwardingServlet.java]: we could if we are worried about using up connection slots on the broker side. I was mostly worried about having the cancellation queue drain fast.

File: server/src/main/java/io/druid/server/AsyncQueryForwardingServlet.java (hunk 2 above)
Comment (xvrl) [inline server/src/main/java/io/druid/server/AsyncQueryForwardingServlet.java]: requests only get queued of all connections to that host are busy. the number of connections by default is already 200

Comment (nishantmonu51) [inline server/src/main/java/io/druid/server/AsyncQueryForwardingServlet.java]: ok, would still be nice to make them configurable, I think.

**Comment by nishantmonu51** [comment]
:+1: after travis

**Comment by drcrallen** [comment]
```
Tests run: 5, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 34.966 sec <<< FAILURE! - in io.druid.indexing.worker.WorkerTaskMonitorTest
testRestartCleansOldStatus(io.druid.indexing.worker.WorkerTaskMonitorTest)  Time elapsed: 31.345 sec  <<< ERROR!
java.lang.Exception: test timed out after 30000 milliseconds
	at sun.misc.Unsafe.park(Native Method)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:186)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:834)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedInterruptibly(AbstractQueuedSynchronizer.java:994)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1303)
	at java.util.concurrent.CountDownLatch.await(CountDownLatch.java:236)
	at io.druid.indexing.worker.WorkerTaskMonitor.stop(WorkerTaskMonitor.java:260)
	at io.druid.indexing.worker.WorkerTaskMonitorTest.testRestartCleansOldStatus(WorkerTaskMonitorTest.java:331)
testRestartCleansOldStatus(io.druid.indexing.worker.WorkerTaskMonitorTest)  Time elapsed: 31.345 sec  <<< ERROR!
java.lang.IllegalStateException: not started
	at com.google.common.base.Preconditions.checkState(Preconditions.java:176)
	at io.druid.indexing.worker.WorkerTaskMonitor.stop(WorkerTaskMonitor.java:250)
	at io.druid.indexing.worker.WorkerTaskMonitorTest.tearDown(WorkerTaskMonitorTest.java:205)
```

File: server/src/main/java/io/druid/server/AsyncQueryForwardingServlet.java [hunk 3]
```diff
@@ -69,6 +69,9 @@
   private static final String QUERY_ATTRIBUTE = "io.druid.proxy.query";
   private static final String OBJECTMAPPER_ATTRIBUTE = "io.druid.proxy.objectMapper";
 
+  private static final int CANCELLATION_TIMEOUT = 500;
```
Comment (drcrallen) [inline server/src/main/java/io/druid/server/AsyncQueryForwardingServlet.java]: Suggest adding `_MS` to name to indicate time unit.

**Comment by drcrallen** [comment]
Timeout is set twice. if I'm understanding the javadocs correctly the init() setting is for simply establishing a connection, while the setting in the call is for the request/response of that specific request.

Anyways, looks good :+1:

**Pull request merged**